magma/backend/mlir/hardware_module.py

```diff
@@ -789,20 +789,25 @@ def visit_module(self, module: ModuleWrapper) -> bool:
         if isinstance(module.module, MagmaInstanceWrapper):
             return self.visit_instance_wrapper(module)
 
-    def visit(self, module: MagmaModuleLike):
-        if module in self._visited:
-            raise RuntimeError(f"Can not re-visit module")
-        self._visited.add(module)
-        for predecessor in self._graph.predecessors(module):
+    def _process_magma_module(self, magma_module: MagmaModuleLike):
+        if not self._graph.has_node(magma_module):
+            return
+        for predecessor in self._graph.predecessors(magma_module):
             if predecessor in self._visited:
                 continue
             self.visit(predecessor)
-        for src, _, data in self._graph.in_edges(module, data=True):
+        for src, _, data in self._graph.in_edges(magma_module, data=True):
             info = data["info"]
             src_port, dst_port = info["src"], info["dst"]
             src_value = self._ctx.get_or_make_mapped_value(src_port)
             self._ctx.set_mapped_value(dst_port, src_value)
-        assert self.visit_module(ModuleWrapper.make(module, self._ctx))
+
+    def visit(self, module: MagmaModuleLike):
+        if module in self._visited:
+            raise RuntimeError(f"Can not re-visit module")
```

Comment:
Would this be an internal error? If so, perhaps we should make it an assertion rather than a runtime error.  That way if a user encounters it it's less likely that they think it's their own fault.

Reply:
Changed it to raise a specific internal error type. I think we should use this pattern for other parts as well, to distinguish external errors from internal errors.